tmva/tmva/inc/TMVA/DNN/Minimizers.h

```diff
@@ -0,0 +1,271 @@
+     fConvergenceCount(0), fTestInterval(0), fLearningRate(0),
+     fMinimumError(1e100)
+{
+   // Nothing to do here.
+}
+//______________________________________________________________________________
+template<typename Architecture_t>
+TGradientDescent<Architecture_t>::TGradientDescent(Scalar_t learningRate,
+                                                   size_t convergenceSteps,
+                                                   size_t testInterval)
+   : fBatchSize(0), fStepCount(0), fConvergenceSteps(convergenceSteps),
+     fConvergenceCount(0), fTestInterval(testInterval), fLearningRate(learningRate),
+     fMinimumError(1e100)
+{
+   // Nothing to do here.
+}
+
+//______________________________________________________________________________
+template<typename Architecture_t>
+template <typename Data_t, typename Net_t>
+    auto TGradientDescent<Architecture_t>::Train(const Data_t & trainingData,
+                                                 size_t nTrainingSamples,
+                                                 const Data_t & testData,
+                                                 size_t nTestSamples,
+                                                 Net_t & net)
+   -> Scalar_t
+{
+   // Reset iteration state.
+   fMinimumError = 1e100;
+   fConvergenceCount = 0;
+   fStepCount = 0;
+
+   // Prepare training data.
+   bool converged = false;
+
+   DataLoader_t<Data_t> trainLoader(trainingData, nTrainingSamples,
+                                    net.GetBatchSize(),
+                                    net.GetInputWidth(), net.GetOutputWidth());
+   auto testNet = net.CreateClone(nTestSamples);
+   DataLoader_t<Data_t> testLoader(testData, nTestSamples,
+                                   testNet.GetBatchSize(),
+                                   testNet.GetInputWidth(), net.GetOutputWidth());
+
+   while (!converged)
+   {
+      for (auto b : trainLoader) {
+         // Perform minimization step.
+         auto inputMatrix  = b.GetInput();
+         auto outputMatrix = b.GetOutput();
+         Step(net, inputMatrix, outputMatrix);
+      }
+
+      // Compute test error.
+      if ((fStepCount % fTestInterval) == 0) {
+         auto b = *testLoader.begin();
+         auto inputMatrix  = b.GetInput();
+         auto outputMatrix = b.GetOutput();
+
+         Scalar_t loss = testNet.Loss(inputMatrix, outputMatrix);
+         std::cout << fStepCount << ": " << loss << std::endl;
+         converged = HasConverged();
+      }
+      fStepCount++;
+   }
+   return fMinimumError;
+}
+
+//______________________________________________________________________________
+template<typename Architecture_t>
+    template <typename Net_t>
+    auto inline TGradientDescent<Architecture_t>::Step(Net_t & net,
+                                                       Matrix_t &input,
+                                                       const Matrix_t &output)
+    -> Scalar_t
```

Comment:
I think it would help the readability of the code (there are likely 'many' Scalar_t) if you were to use Architecture_t::Scalar_t even in this notation.